File: server/test/e2e/TestLogin.ts
import {browser} from "protractor";
import e2eHelp from "./e2e-help";
import TribeListPage from "./page-objects/TribeListPage";

const config = require("../../config/config");
const hostName = `http://${config.publicHost}:${config.port}`;

export default {
    async login() {
        await browser.get(`${hostName}/test-login?username=${e2eHelp.userEmail}&password="pw"`);
        await TribeListPage.waitForTribeListPage();
    }
}

File: server/src/endToEndTest/kotlin/com/zegreatrob/coupling/server/e2e/TestLogin.kt
package com.zegreatrob.coupling.server.e2e

import com.zegreatrob.coupling.server.e2e.external.webdriverio.BrowserSyntax
import com.zegreatrob.coupling.server.e2e.external.webdriverio.WebdriverBrowser

object TestLogin : BrowserSyntax {
    suspend fun login(userEmail: String) {

        for (attempt in 1..3) {
            try {
                tryLogin(userEmail)
                break
            } catch (throwable: Throwable) {
                println("Failed login attempt $attempt")
                if (attempt == 3) throw throwable
                WebdriverBrowser.setUrl("/")
            }
        }
        clearLogs()
    }

    private suspend fun clearLogs() {
        WebdriverBrowser.getLogs()
    }

    private suspend fun tryLogin(userEmail: String) {
        WebdriverBrowser.setUrl("${Config.publicUrl}/test-login?username=${userEmail}&password=pw")
        TribeListPage.waitForPage()
    }
}